tools/get_dicom.pl

```diff
@@ -0,0 +1,275 @@
+
+#---------------------------------------#
+# Read prod file to get DB credentials  #
+#---------------------------------------#
+{ package Settings; do "$ENV{LORIS_CONFIG}/.loris_mri/$profile" }
+if(!@Settings::db) {
+    die "No database settings in file $ENV{LORIS_CONFIG}/.loris_mri/$profile\n";
+}
+my $dbh = &NeuroDB::DBI::connect_to_db(@Settings::db);
+
+#-------------------------------------------#
+# Fetch tarchiveLibraryDir from the config  #
+#-------------------------------------------#
+my $sth = $dbh->prepare(
+     "SELECT Value "
+   . "FROM Config c "
+   . "JOIN ConfigSettings cs ON (cs.ID=c.ConfigID) "
+   . "WHERE cs.Name = ?"
+);
+$sth->execute('tarchiveLibraryDir');
+
+my($tarchiveLibraryDir) = $sth->fetchrow_array;
+
+#------------------------------------------------------------#
+# Get the list of tarchives for which the patient names      #
+# match the list of supplied patterns                        #
+#------------------------------------------------------------#
+my $stmt = "SELECT c.pscid, s.visit_label, t.dateacquired, t.archivelocation, t.tarchiveid, t.patientname "
+         . "FROM tarchive t "
+         . "JOIN session s ON (t.sessionid=s.id) "
+         . "JOIN candidate c ON (c.candid=s.candid) ";
+if(@patientNames) {
+	my @where = map { "t.patientname LIKE ?" } @patientNames;
+	$stmt .= sprintf("WHERE %s", join("OR", @where));
+}
+$sth = $dbh->prepare($stmt);
+
+if(@patientNames) {
+    for(my $p=0; $p<@patientNames; $p++) {
+        $sth->bind_param($p+1, $patientNames[$p]);
+    }
+}
+$sth->execute();
+
+#----------------------------------------------------------#
+# Create the directory where DICOM files will be extracted #
+#----------------------------------------------------------#
+my $tmpExtractDir = sprintf("%s/%s.%d", $tmpExtractBaseDir, basename($0), $$);
+mkdir $tmpExtractDir or die "Cannot create directory $tmpExtractDir: $!\n";
+
+#---------------------------------------------------------#
+# Process each tarchive and extract the DICOMs associated #
+# to the scan types of interest                           #
+#---------------------------------------------------------#
+my @outDirs = ();
+foreach my $tarchiveRowRef (@{ $sth->fetchall_arrayref }) {
+    my($pscid, $visitLabel, $dateAcquired, $archiveLocation, $tarchiveId) = @$tarchiveRowRef;
+    
+    my($innerTar) = "$archiveLocation.gz" =~ /_(ImagingUpload.*.tar.gz)$/;
+    
+    # Extract only the .tar.gz archive from the main archive (ignore the 
+    # meta data and log file)
+    print "Extracting $innerTar in $tmpExtractDir...";
+    system("tar xf $tarchiveLibraryDir/$archiveLocation -C $tmpExtractDir $innerTar") == 0 
+      or die "Extraction of $innerTar from $tarchiveLibraryDir/$archiveLocation in directory $tmpExtractDir failed: $?";
+    print "done\n";
+    
+    # Fetch all the MINC files in the archive whose acquisition protocols
+    # match the scan types of interest
+    $stmt = "SELECT f.file, tf.filename "
```

Comment:
Not sure I understand the goal of that query. The way I understand it is that it selects ALL DICOM files present in the tarchive_files table for a given tarchiveID (regardless of the scan type you specify so why not directly joining the tarchive_files based on SeriesUID and EchoTime instead to get the proper list of DICOM files?).

Basically:
```
SELECT f.File, tf.filename
FROM files f
JOIN mri_scan_type mst ON (mst.ID=f.AcquisitionProtocolID)
JOIN tarchive_series ts ON (f.SeriesUID=ts.SeriesUID AND f.EchoTime*1000=ts.EchoTime)
JOIN tarchive_files tf USING (TarchiveSeriesID)
WHERE TarchiveSource = ?
```
Then keep your where for the scanType to append to that query.

FYI, have to multiply EchoTime in `files` table by 1000 so that it is the same unit as what is stored in `tarchive_series`...

Using the query above, you get directly all the DICOM files you need to archive and no need for further queries to filter them out :)

Reply:
also, relying on the SeriesUID/EchoTime combination is more robust than relying on seriesDescription and scan number

Reply:
Catch of the year! That was really not working as it should. Thanks and fixed!